### PR TITLE
ISSUE-688 # Propagate exceptions from java api operations

### LINKED ISSUE #688: Exceptions thrown from classes executed by the java api are not propagated
When a java api operation throws an exception, the exception message and stacktrace are not shown as part of the scenario stacktrace, making it difficult to debug.

**To reproduce:**
- Start from a new branch
- Edit `org.jsmart.zerocode.core.AddService` and replace `return 30` in the 'anInteger' method by `throw new RuntimeException("INTENTIONAL EXCEPTION")`
- Optionally, edit the resource file `unit_test_files/java_apis/02_test_json_java_service_method_no_param.json` and remove `"loop": 5,`
- Run the test in package `org.jsmart.zerocode.core.javamethod`

**Expected:** Some information about the RuntimeException `INTENTIONAL EXCEPTION`, and its stacktrace should be shown
**Actual:** It is ommited. This is the stacktrace:

```
2024-10-20 11:55:44,184 [main] ERROR org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl - Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService. Exception - java.lang.reflect.InvocationTargetException
2024-10-20 11:55:44,184 [main] ERROR org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl - Exception - java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
2024-10-20 11:55:44,185 [main] ERROR org.jsmart.zerocode.core.engine.executor.ApiServiceExecutorImpl - Java method execution exception - java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
	at org.jsmart.zerocode.core.engine.executor.ApiServiceExecutorImpl.executeJavaOperation(ApiServiceExecutorImpl.java:53)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeApi(ZeroCodeMultiStepsScenarioRunnerImpl.java:471)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeRetry(ZeroCodeMultiStepsScenarioRunnerImpl.java:256)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeRetryWithSteps(ZeroCodeMultiStepsScenarioRunnerImpl.java:195)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeSteps(ZeroCodeMultiStepsScenarioRunnerImpl.java:177)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.runScenario(ZeroCodeMultiStepsScenarioRunnerImpl.java:140)
(...)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.execute(JavaMethodExecutorImpl.java:63)
	at org.jsmart.zerocode.core.engine.executor.ApiServiceExecutorImpl.executeJavaOperation(ApiServiceExecutorImpl.java:50)
	... 22 more
Caused by: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.executeWithParams(JavaMethodExecutorImpl.java:114)
	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.execute(JavaMethodExecutorImpl.java:47)
	... 23 more
2024-10-20 11:55:44,186 [main] ERROR org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl - ###Exception while executing a step in the zerocode dsl.
2024-10-20 11:55:44,188 [main] ERROR org.jsmart.zerocode.core.runner.StepNotificationHandler - Exception occurred while executing Scenario:[Given_When_Then-Flow name For Java Service], --> Step:[javaMethodExecution], Details: java.lang.RuntimeException: ZeroCode Step execution failed. Details:java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
2024-10-20 11:55:44,193[main] org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl -m
```

### PR DESCRIPTION
**Pull request by javiertuya**
# Propagate exceptions from java api operations

## Fixes Issue
- [x] Which issue or ticket was(will be) fixed by this PR? Fix #688 

PR Branch
**https://github.com/javiertuya/zerocode/tree/688-exceptions-java-api**

## Motivation and Context

When executing a java method that is invoked by the java api as an operation, any exception in the java method should be shown in the stacktrace to facilitate debugging. The executor exception handler should include this exception

## Checklist:

* [x] New Unit tests were added
  * [ ] Covered in existing Unit tests

* [ ] Integration tests were added
  * [ ] Covered in existing Integration tests

* [x] Test names are meaningful

* [x] Feature manually tested and outcome is successful

* [x] PR doesn't break any of the earlier features for end users
  * [ ] WARNING! This might break one or more earlier earlier features, hence left a comment tagging all reviewrs

* [x] Branch build passed in CI

* [x] No 'package.*' in the imports

* [ ] Relevant DOcumentation page added or updated with clear instructions and examples for the end user
  * [x] Not applicable. This was only a code refactor change, no functional or behaviourial changes were introduced

* [ ] Http test added to `http-testing` module(if applicable) ?
  * [x] Not applicable. The changes did not affect HTTP automation flow

* [ ] Kafka test added to `kafka-testing` module(if applicable) ?
  * [x] Not applicable. The changes did not affect Kafka automation flow

### CODE REVIEW

**Review by authorjapps — APPROVED** [review]

**Comment by javiertuya** [comment]
@authorjapps Now that I've collaborator access, I have a few questions about the PR workflow on this repo:
- Should I merge PRs submitted by me once they get the approval (like this), or let the reviewers do it?
- Just after I create a PR, should I assign you as reviewer, place a mention or nothing?

**Comment by authorjapps** [comment]
> @authorjapps Now that I've collaborator access, I have a few questions about the PR workflow on this repo:
> 
> * Should I merge PRs submitted by me once they get the approval (like this), or let the reviewers do it?
> * Just after I create a PR, should I assign you as reviewer, place a mention or nothing?

@javiertuya, Thanks for asking.
You can merge it, if nothing is outstanding from any reviewers.

In this case, yes, go ahead and merge it 👍  
and then, just watch the master build to go Green automatically.

**Pull request merged**
